File: app/domain/type/fizz_buzz_type.go
package types

import "fizzbuzz/domain/model"

// FizzBuzzType FizzBuzzタイプインターフェース（ポリモーフィズム）
type FizzBuzzType interface {
	Generate(number int) *model.FizzBuzzValue
}

// FizzBuzzTypeBase 共通メソッドを提供する基底構造体
type FizzBuzzTypeBase struct{}

func (f *FizzBuzzTypeBase) IsFizz(number int) bool {
	return number%3 == 0
}

func (f *FizzBuzzTypeBase) IsBuzz(number int) bool {
	return number%5 == 0
}

// CreateFizzBuzzType ファクトリメソッド
func CreateFizzBuzzType(typeNum int) FizzBuzzType {
	switch typeNum {
	case 1:
		return &FizzBuzzType01{}
	case 2:
		return &FizzBuzzType02{}
	case 3:
		return &FizzBuzzType03{}
	default:
		panic("不正なタイプです")
	}
}

File: app/lib/domain/type/fizz_buzz_type.js
class FizzBuzzType {
  isFizz(number) {
    return number % 3 === 0;
  }

  isBuzz(number) {
    return number % 5 === 0;
  }

  static create(type) {
    switch (type) {
      case 1:
        return new (require('./fizz_buzz_type_01'))();
      case 2:
        return new (require('./fizz_buzz_type_02'))();
      case 3:
        return new (require('./fizz_buzz_type_03'))();
      default:
        return new (require('./fizz_buzz_type_not_defined'))();
    }
  }
}

module.exports = FizzBuzzType;
